text/0082-dynamic-size-region.md

```diff
@@ -0,0 +1,203 @@
+# Dynamic size region
+
+- RFC PR: https://github.com/tikv/rfcs/pull/0082
+- Tracking Issue: https://github.com/tikv/tikv/issues/11515
+
+## Summary
+
+Make the size of a region dynamic, and only set a upper limit.
+
+This is the first step that we try to support PiB scale cluster.
+
+## Motivation
+
+We have observed a lot of regressions when the count of regions increases. The major drawback
+comes from three aspects:
+
+1. Transactions need to access more regions, which lead to too many RPCs and have high latency;
+2. A node needs to drive more regions, which can have high resource usages and hurt performance;
+3. Tools and service like CDC/BR/GC need to loop over all regions, which is slow.
+
+We have hibernate raft (#23) to get around the most significant problem 2. But it's not always
+working as expected:
+
+- Workload with random access can still activate a lot of regions;
+- Hibernated region needs extra time to renew its lease, which lead to high tail latency;
+- It complicates the design of CDC/BR, which requires access all replicas instead of just leader now;
+- GC will still wake up all regions and cause periodical high usage.
+
+There are two source of region creations: size/keys split and table split. Table split is disabled by
+default in TiKV. Even if it splits a lot of regions, small one will be merged later.
+
+Size/keys split contributes the most count of regions. The current size of a region is 96MiB by default,
+and the key count is 960,000. They are set based on an assumption that region is the concurrent unit of
+requests. Too large can lead to smaller concurrency and high latency. Since keys is related to the size,
+we only discuss region size here.
+
+## Detailed design
+
+To reduce the count of regions, we can set the size of region to a larger value. More precisely, instead
+of 96MiB, we set the max region size to 10GiB (before compression). To get around the problem of
+concurrency, we introduce new concept bucket that is smaller ranges within a region. To make hot spot be
+recognized and scheduled quickly, we choose the make the size of a region dynamic.
+
+![dynamic size](../media/dynamic-size.png)
+
+### Dynamic size
+
+The hotter a region is, the smaller its size becomes. To make it simplified, we choose 512MiB for hot regions
+and 10GiB for cold regions. So there are two types of split, hot spot split and general size split. Hot spot
+split is triggered by PD. General size split is triggered by TiKV itself.
+
+10GiB is just an example, it's allowed to change to a bigger or smaller value.
+
+### Bucket
```

Comment:
Since we measure the read/write load at bucket level, which means it's about 128MB size. How can we further split a hot bucket? Today we have load-based split, which may cut the small region by the access pattern.

Reply:
I don't have a clear idea now. I will leave it as an unresolved question. Maybe we can split the bucket further. For example, besides controlling the size of a bucket, also require the minimal count of a bucket. So even a region becomes small, it can still be split further.